Enable time-dependent min/max refinement

jperryhouts opened the pull request:
I'm using a case right now where I'd like the minimum and maximum refinement levels to change with time, which was apparently not previously supported.

gassmoeller commented:
Hi Jonathan,
I see what you are trying to do, and I think it is a nice improvement, however I would prefer a slightly different approach. Instead of removing the 'const' classification and changing the interface, you could add an `update()` function to the interface that gets called once per time step and that updates the time of the function in your specific plugin. It would need an default implementation in the interface that does nothing to not break compatibility with other plugins, but it would be more similar to our other interfaces (e.g. VelocityBoundaryConditions) and it would allow other plugins to do other special things (if they want to). The tricky point would be to find the right place for the call to the update function, but it would likely be in the `Simulator<dim>::start_timestep()` function where all the other update functions are called. 
Would that be a possible approach for you, or is there something that would not work that way?

jperryhouts commented:
Good idea.

One question: All the other plugin interfaces which have an `update()` function are  `const unique_ptr`'s in the Simulator, but the `mesh_refinement_manager` is just a regular old object. This is one of those parts of C++ that I've never really gotten a good grasp on, what's the rationale for one vs the other? And should the `mesh_refinement_manager` be a pointer for consistency, or can it stay as it is?

bangerth commented:
No. If there can only be once plugin object at a time, we use a `unique_ptr`. An example is the material model. But if multiple plugin objects can exist at the same time -- think postprocessors, or mesh refinement objects -- then there is typically a manager object that internally stores pointers to the plugin objects. The `update()` function of the manager would then typically just call the `update()` functions of the objects to which the manager stores pointers.

jperryhouts commented:
Ah, makes sense. Thanks for clarifying! So, does this PR (as amended) correctly implement that model?

bangerth commented:
/run-tests

bangerth commented on include/aspect/mesh_refinement/interface.h:
```diff
@@ -89,6 +89,19 @@ namespace aspect
 
         /**
+         * A function that is called once at the beginning of the refinement
+         * process. The default implementation of the function does nothing,
```
That comment is not quite correct (or at least shouldn't be correct). The function should be called once per time step.

bangerth commented:
Other than the one comment, this actually looks good. Can you
* add an entry to `changes.h`
* maybe update the doc string for the two changed refinement models to make clear that the function expression one specifies in the input file may also reference the time (and that the time is counted in seconds or, if appropriate, in years)?
* do you think it would be possible clone the existing tests for these refinement plugins to also have something that tests the time dependence?

jperryhouts commented:
> this actually looks good

Um, thanks? I'd attribute this one to judicious copy-pasting. :)

I addressed those points. I only added one test (it tests the `minimum refinement function` plugin, but not the `maximum refinement function` one), because running multi-timestep tests is a bit slow. Let me know if you think it needs a test for the other case.

jperryhouts commented:
Is something up with the tester? The results all say "0 tests failed"

gassmoeller commented:
Sometimes the tester has problems uploading the results. The tests are passing though, so lets merge anyway. Thanks. :+1:

jperryhouts commented:
Cool, thanks!
